src/QsCompiler/Transformations/CallGraphWalker.cs

```diff
@@ -0,0 +1,344 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.Quantum.QsCompiler.DataTypes;
+using Microsoft.Quantum.QsCompiler.SyntaxTokens;
+using Microsoft.Quantum.QsCompiler.SyntaxTree;
+using Microsoft.Quantum.QsCompiler.Transformations.Core;
+
+
+namespace Microsoft.Quantum.QsCompiler.Transformations.CallGraphWalker
+{
+    using ExpressionKind = QsExpressionKind<TypedExpression, Identifier, ResolvedType>;
+    using ResolvedTypeKind = QsTypeKind<ResolvedType, UserDefinedType, QsTypeParameter, CallableInformation>;
+
+    /// Class used to track call graph of a compilation
+    public class CallGraph
+    {
+        public struct CallGraphEdge
+        {
+            public ImmutableDictionary<Tuple<QsQualifiedName, NonNullable<string>>, ResolvedType> ParamResolutions;
+        }
+
+        public struct CallGraphNode
+        {
+            public QsQualifiedName CallableName;
+            public QsSpecializationKind Kind;
+            public QsNullable<ImmutableArray<ResolvedType>> TypeArgs;
+        }
+
+        private Dictionary<CallGraphNode, Dictionary<CallGraphNode, ImmutableArray<CallGraphEdge>>> _Dependencies =
```

Comment:
>Dictionary<CallGraphNode, Dictionary<CallGraphNode, ImmutableArray<CallGraphEdge>>> [](start = 16, length = 83)

Do I understand it correctly, that this is a dictionary mapping source node to pairs of target node and information about the edge between source node and target node? Would be great to add a comment stating it. This is a non-trivial place for understanding.